woocommerce/woocommerce-ios
# Order Creation: Add discard changes prompt to back navigation

Closes: #5815
⚠️ Based on #6397; make sure that PR is merged first. ⚠️ 

## Description

During Order Creation, we want a "discard changes" prompt whenever the order has changed and the new order screen is being dismissed. In #6381 we added a "discard changes" prompt when the new order view is presented modally (when the `splitViewInOrdersTab` feature flag is enabled). This adds the same prompt when the view is pushed onto the navigation stack (when that flag is disabled), and the user taps the Back button or swipes to navigate back.

It also ensures the order is deleted remotely when "discard changes" is selected, regardless of how the new order screen is presented. (This is particularly useful to clean up when the store has < WC 6.3, and new orders are synced with `pending` status rather than `auto-draft`.)

## Changes

* Adds the discard changes prompt to `NewOrderHostingController` when the controller is being popped from the back button or swipe gesture.
* Adds a `discardOrder` method to `NewOrderViewModel`; this method is called in both cases when the discard action is called in `NewOrderHostingController` (when view is presented modally or non-modally).
* Adds a `discardOrder` method to `OrderSynchronizer`:
   * In the local synchronizer, this is a no-op (the local order will be discarded when navigating back).
   * In the remote synchronizer, does nothing if the order is local (same as above) and deletes the order remotely if it has been synced already.
* Adds unit tests for discarding orders.

## Testing

1. Set the `splitViewInOrdersTab` feature flag to `false` in `DefaultFeatureFlagService`.
2. Build and run the app on a real device (not a simulator).
3. Navigate to the Orders tab and create a new order.
4. Tap the back button or swipe back, and confirm the view is dismissed without any prompt.
5. Create a new order.
6. Make any changes to the order.
7. Tap the back button or swipe back, and confirm the "discard changes" prompt appears.
8. On a store with < WC 6.3, make sure the order is synced remotely (add a product, shipping, or fee) and then discard the order. Confirm it doesn't appear in the order list.

You can also repeat the test on a store with < WC 6.3, with the `splitViewInOrdersTab` feature flag enabled, and confirm the discarded order does not appear in the order list.

## Screenshots


https://user-images.githubusercontent.com/8658164/157945990-a1d36192-016b-4103-a4f5-49c410b53c23.MP4



## Submitter Checklist

Update release notes:
